### PR TITLE
Add import name resolution to `Context`

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Right now, we make heavy use of `resolve_call_path`, which takes an expression as input and resolves it to a fully qualified import name, if possible.

For example, given:
```python
from sys import version_info as python_version
print(python_version)
```

...then `resolve_call_path(${python_version})` would resolve to `sys.version_info`.

This PR implements the reverse lookup: given an import name, can we find a binding in the current scope that maps to that import name? E.g., if we want to insert a reference to `sys.version`, can we write a function that returns `python_version`?

This method is implemented here as `resolve_binding`. It builds on logic that already existed in `crates/ruff/src/rules/pylint/rules/sys_exit_alias.rs`, but (1) generalizes it such that any module can tap into it, and (2) fixes at least one bug. I've also modified an additional rule (`crates/ruff/src/rules/pyupgrade/rules/lru_cache_with_maxsize_none.rs`) to make use of it.

### CODE REVIEW

File: crates/ruff/resources/test/fixtures/pylint/sys_exit_alias_3.py
Comment (charliermarsh) [inline crates/ruff/resources/test/fixtures/pylint/sys_exit_alias_3.py:16]: The existing version of `resolve_binding` in `sys_exit_alias` didn't take into account that bindings might be overridden. We now avoid trying to fix this (since the `exit` in `from sys import exit` (out-of-view, on Line 1) is overridden here).

File: crates/ruff_python_ast/src/context.rs
Comment (charliermarsh) [inline crates/ruff_python_ast/src/context.rs:235]: This whole thing has the potential to be pretty slow.

We have to iterate over all bindings in scope (to find candidates), then iterate over the scopes we passed over to ensure that the candidates weren't overridden at a lower level. I'm sure we could make this much faster, but it'd require different bookkeeping.

**Comment by JonathanPlasse** [comment]
Glad to see my code reused and improved :smiley:

**Comment by charliermarsh** [comment]
👏 👏 👏

File: crates/ruff_python_ast/src/context.rs
Comment (MichaReiser) [reply crates/ruff_python_ast/src/context.rs:229]: I think what we have here is "fine" for now but we should invest into building a full semantic model as part of the parsing or syntax traversal, and then use it to answer these kind of questions.

File: crates/ruff_python_ast/src/context.rs
Comment (MichaReiser) [inline crates/ruff_python_ast/src/context.rs:220]: Nit: You could avoid allocating in some cases by returning `Option<Cow<str>>`, but I don't know if its worth here or if it would be better to return an `enum` instead that then has a method to get the full qualified name of the local symbol referencing the import.

**Review by MichaReiser — APPROVED** [review]
This looks good to me to unblock some work, but I think we want to take a more holistic approach to solve semantic analysis / symbol table in the future to better balance performance with the full power of a symbol table and to enable cross file analysis. 

I think our ultimate goal should be to have a semantic model that exposes rich information about symbols and lets us:

* resolve the symbol for a reference
* get all references of a symbol
* get all symbols for a scope
* get all declarations/definitions of a symbol. 
* get all members of a symbol (class members)
* get all exports/imports of a module (similar to getting all members of a symbol)

Similar to what the semantic model of [oxc](https://github.com/Boshen/oxc/tree/main/crates/oxc_semantic), [rome](https://github.com/rome/tools/blob/main/crates/rome_js_semantic/src/semantic_model/model.rs), or [TypeScript](https://basarat.gitbook.io/typescript/overview/binder) offer. 

Our own parser would allow us to build the semantic model (or the scope table at least) while parsing or building the AST. But it may still be required to split ruff into multiple passes:

* syntax lints
* semantic lints
* project/solution lints

## Terminology 

I noticed is that we have a different understanding of the terms *symbol* and *binding* that I think is worth clarifying. Let me explain my understanding and then try to align on a shared terminology. 

I'll use the following code for  my explanation:

```python
import sys

class Test:
	class_field = 1
	
	def __init__(self):
		self.a = 10

	def class_method():
		pass

	def method(self):
		pass

def f():
	pass

a = 10
a = a + 2

test = new Test()
test.a = 2
```

### Symbols

A symbol is the unique combination of a "name" and its scope. For example, `a` in `a + 2` references the variable `a` from the local scope, whereas `test.a` references the class instance symbol `a` of `Test`. 

Every symbol has a name that uniquely identifies it in the current scope: Examples are `a`, `test`, `Test`, `sys`. 

A symbol can have a full qualified name. For example: the full qualified name of the member `exit` on `sys` is `sys.exit`.

### Definition/Declaration
A declaration is the expression or statement that *creates* the new symbol. `class Test` is the declaration for the `Test` symbol, or `a = 10` declares the variable `a`. 

A symbol can have zero to 1 or many definition/declarations. 

* 0: global or undeclared symbol
* 1: Most common: Symbol with a declaration in source code or a definition in a typing declaration file.
* Many: 
	* symbol defined in a typing declaration file AND declared in a python file. 

### References / Uses
Read or write access to a symbol after its declaration. This can be a function call (lookup of the function), variable or member assignments, or normal variables. 

Examples:
* `a = a + 2`: References the `a` symbol
* `new Test()`: References the `Test` and, implicitly, the `__init__` symbol
* `test.a = 3`: References the `test` and the `Test.a` symbol

It can be useful to track whether it's a read or write a reference, e.g., to recommend renaming a supposedly constant to a non-constant if there are any write references.

### Binding
The part I struggle most with explaining and have only found little reference is *Binding*. 

I think that a *binding* is the combination of a *Symbol* with its value (or metadata). In our case the *Binding* is the symbol table entry that stores the mapping from symbol to its declarations and references. 

For example, `a` is a binding. It has one declaration `a = 10`, and two references: `a`, and 
a = a + 2`

### Scope
Implements the encapsulation of symbols: symbols are only visible to child scopes and shadowing overrides variables from the outer scope. 

### Symbol Table
Stores the symbols of a program. A question here is if symbols with members (e.g. classes) have their own symbol table/ implement symbol table too, or if that's implemented on top of scopes. 

### Semantic model
Contaisn the symbol table but may also give access to type information. 

Stores the symbols for a program

**Comment by MichaReiser** [comment]
> This would introduce two symbols (exit and foo). I think they'd have the same "fully-qualified name"? And they are references to the same underlying "thing" in sys, though I'm not certain what that thing should be called. I think it's a symbol in sys.

This is a great question, and I don't have a particularly good answer for it. It would be interesting to look at other tools.

I understand that there are multiple symbols at play. 
* One `external` symbol `sys` for the module
* The module `sys` has one member `exit` 
* The local symbol `exit` is an alias for the `exit` member of `sys`. 
* The local symbol `foo` is an alias for the `exit` member of `sys`.

So yes, both `exit` and `foo` point to the same external symbol. But I'm not sure how to best represent this relationship. 

> In this case, it's just one symbol (a) even though it's rebound, right?

Yes, depending on the definition, `a` as two or three references: one read, and one or two writes.

**Comment by charliermarsh** [comment]
In this framing, would you agree that the _key_ to the `scope.values` hash table is the "symbol", and the value is the "binding" (`Binding`)?

File: crates/ruff_python_ast/src/context.rs
Comment (charliermarsh) [reply crates/ruff_python_ast/src/context.rs:220]: It's not `pub(crate)` though, we need to access this from `ruff` :)

File: crates/ruff_python_ast/src/context.rs
Comment (charliermarsh) [reply crates/ruff_python_ast/src/context.rs:223]: `binding` iterates over (name, `BindingId`) pairs, so we still have to do this lookup I think?

**Comment by MichaReiser** [comment]
> In this framing, would you agree that the _key_ to the `scope.values` hash table is the "symbol", and the value is the "binding" (`Binding`)?

Not sure. In my mind the binding is the association of a "name" (Symbol) to its location (if you emit bytecode) or metadata (static analyzer). But I think that's a detail and calling the metadata `Binding` is fine (especially if it has a back reference to the symbol)

**Pull request merged**
